### DISCUSSION TITLE
how to set a printcolumn from multi value, like deployment READY status

### DISCUSSION THREAD

**Discussion by nfyxhan** [kubernetes-sigs/kubebuilder#2285] [Support]
```golang
type FooSpec struct {
       Ready int `json:"ready"`
       Total int `json:"total"`
}
// +kubebuilder:object:root=true
// +kubebuilder:printcolumn:name="ready",type=string,JSONPath=`.ready,.total` // todo
type Foo struct {
      metav1.TypeMeta   `json:",inline"`
      metav1.ObjectMeta `json:"metadata,omitempty"`
  
      Spec   FooSpec   `json:"spec,omitempty"`
      Status FooStatus `json:"status,omitempty"`
 }
```
I wander how to set the printcolum to make ready column like this.
```
# kubectl get foo
NAME       READY
foo1         1/1
foo2         0/1
```

**Comment by camilamacedo86** [comment]
To achieve the desired output where the READY column displays as a fraction (e.g., 1/1), you'd need to use the JSONPath feature in kubebuilder:printcolumn.

Here's how you can set it up for a single field:

```go
// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".spec.ready"
```
For your desired output, you may have to look into extending kubectl or writing a custom script to fetch and format the resources in your desired way.

If the core Kubernetes project or the controller-tools project (which handles the generation of CRD manifests from the kubebuilder annotations) so you might want check the documentation and raise issues for their projects as well as raise questions in the controller-runtime slack channel.

I hope that it answered your question.
I am closing this one as answered, however, please feel free to re-open if you need to do so.

**Comment by camilamacedo86** [comment]
Following an example if the above explanation is not clear enough:

```go
type FooSpec struct {
       Ready int `json:"ready"`
       Total int `json:"total"`
}

// +kubebuilder:object:root=true
// +kubebuilder:printcolumn:name="READY",type=string,JSONPath=`{.spec.ready}/{.spec.total}`
type Foo struct {
      metav1.TypeMeta   `json:",inline"`
      metav1.ObjectMeta `json:"metadata,omitempty"`
  
      Spec   FooSpec   `json:"spec,omitempty"`
      Status FooStatus `json:"status,omitempty"`
}
```

Please, let us know if that is not what you are looking for. If not please, could you please clarify what is missing?

**Reply by nfyxhan** [reply]
Invalid value: "{.spec.ready}/{.spec.total}": must be a simple json path starting with .

**Reply by Arakos** [reply]
for people finding this issue: support for the specified behavior is not yet supported, discussions on that are happening in 
https://github.com/kubernetes-sigs/controller-tools/issues/1051#issuecomment-2337392003

**Comment by mattwelke** [comment]
As noted in https://github.com/kubernetes-sigs/kubebuilder/discussions/2285#discussioncomment-11508057, this seems to not be available yet (the ability to specify a complex expression that uses one or more JSON paths to reference more than one part of the CR).

I think I was able to come up with a workaround though. I add a string to my CRD's status that can be used to convey this information. Then, I point the printcolumn annotation to that single field in the status.

type:

```go
type FooSpec struct {
	Total int `json:"total,omitempty"`
}

type FooStatus struct {
	Ready string `json:"ready"`
}

// +kubebuilder:object:root=true
// +kubebuilder:subresource:status
// +kubebuilder:printcolumn:name="READY",type=string,JSONPath=`.status.ready`
type Foo struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   FooSpec   `json:"spec,omitempty"`
	Status FooStatus `json:"status,omitempty"`
}
```

controller:

```go
func (r *FooReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	// Log the request for debugging purposes
	log := log.FromContext(ctx)
	log.Info("Reconciling Foo", "namespace", req.Namespace, "name", req.Name)

	// Fetch the Foo resource
	var foo examplesv1.Foo
	if err := r.Get(ctx, req.NamespacedName, &foo); err != nil {
		if apierrors.IsNotFound(err) {
			// Resource not found, nothing to reconcile
			log.Info("Foo resource not found, ignoring since object must be deleted")
			return ctrl.Result{}, nil
		}
		// Error reading the object
		log.Error(err, "Failed to get Foo resource")
		return ctrl.Result{}, err
	}

	// Update the status - use "1" for the number ready for this test
	foo.Status.Ready = fmt.Sprintf("%s/%d", "1", foo.Spec.Total)

	// Save the updated status back to the cluster
	if err := r.Status().Update(ctx, &foo); err != nil {
		log.Error(err, "Failed to update Foo status")
		return ctrl.Result{}, err
	}

	log.Info("Successfully updated Foo status", "namespace", req.Namespace, "name", req.Name)
	return ctrl.Result{}, nil
}
```

sample CR:

```yaml
apiVersion: examples.example.com/v1
kind: Foo
metadata:
  labels:
    app.kubernetes.io/name: kubebuilder-test-dec9
    app.kubernetes.io/managed-by: kustomize
  name: foo-sample
spec:
  total: 2
```

This causes the string field to be included as is in the status:

```yaml
apiVersion: examples.example.com/v1
kind: Foo
...
status:
  ready: 1/2
```

One thing I liked about this approach is that it seemed so flexible. It seems like any data can be conveyed this way, as long as you can format it into a string.
